fetch.bs

```diff
@@ -5630,6 +5621,81 @@ if the script checks that the URL has the right hostname.
 
 
 
+<h2 id=data-urls><code>data:</code> URLs</h2>
+
+<p>For an informative description of <code>data:</code> URLs see RFC 2397. This section replaces its
+normative processing requirements. [[!RFC2397]]
+
+<p>The <dfn export><code>data:</code> URL processor</dfn> takes a <a for=/>URL</a>
+<var>dataURL</var> and then runs these steps:
+
+<ol>
+ <li><p>Assert: <var>dataURL</var>'s <a for=URL>scheme</a> is "<code>data</code>".
+
+ <li><p>Let <var>input</var> be the result of running the <a>URL serializer</a> on
+ <var>dataURL</var> with the <i>exclude fragment flag</i> set.
+
+ <li><p>Remove the leading "<code>data:</code>" string from <var>input</var>.
+
+ <li><p>Let <var>position</var> point at the start of <var>input</var>.
+
+ <li><p>Let <var>encodedMimeType</var> be the result of
+ <a>collecting a sequence of code points</a> that are not equal to U+002C (,), given
+ <var>position</var>.
+
+ <li><p>If <var>position</var> is past the end of <var>input</var>, then return failure.
+
+ <li><p>Advance <var>position</var> by 1.
+
+ <li><p>Let <var>encodedBody</var> be the remainder of <var>input</var>.
+
+ <li><p>Let <var>mimeTypeBytes</var> be the <a>string percent decoding</a> of
+ <var>encodedMimeType</var>.
+ <!-- Note: implementations leave the percent-encoded bits around. That strikes me as rather broken,
+      but it might be due to web compatibility? -->
+
+ <li><p>Let <var>mimeType</var> be <var>mimeTypeBytes</var> with each byte mapped to a code point of
+ equal value.
+ <!-- Maybe we should define a dedicated function for this. I've also used this for Refresh, IDL
+      needs it for ByteString. Maybe there's other locations? -->
+
+ <li><p><a>Strip leading and trailing ASCII whitespace</a> from <var>mimeType</var>.
+
+ <li><p>Let <var>body</var> be the <a>string percent decoding</a> of <var>encodedBody</var>.
+
+ <li>
+  <p>If <var>mimeType</var> contains an <a>ASCII case-insensitive</a> match for
+  "<code>;base64;</code>" or ends with an <a>ASCII case-insensitive</a> match for
+  "<code>;base64</code>", then:
+
+  <ol>
+   <li><p>Set <var>body</var> to the <span class=XXX>web base64 decoding</span> of <var>body</var>.
+   <!-- web base64 currently takes string input (see window.atob, which defines it for now). I guess
+        we can convert body to a string first by straightforward mapping, but maybe it would make
+        more sense if web base64 decoding operated on bytes? Or was polymorphic? -->
+
+   <li><p>If <var>body</var> is failure, then return failure.
+
+   <li><p>Remove the code point sequence that is an <a>ASCII case-insensitive</a> match for
+   "<code>;base64</code>" from <var>mimeType</var>.
+  </ol>
+
+ <li><p>If <var>mimeType</var> starts with an <a>ASCII case-insensitive</a> match for
+ "<code>charset;</code>", then prepend "<code>text/plain</code>" to <var>mimeType</var>.
```

Comment:
The latter doesn't preserve anything in Chrome and Firefox though. The former does seem to preserve the charset value somehow, but doesn't work at all in Safari. I'm not sure we need to support it.